contrib/glossary/glossary.go

```diff
@@ -0,0 +1,138 @@
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+)
+
+type Config struct {
+	inputFile    string
+	outputFile   string
+	glossaryFile string
+}
+
+type GlossaryEntry struct {
+	Words []string
+	Link  string
+}
+
+type Glossary struct {
+	Items []GlossaryEntry
+}
+
+func loadGlossary(file string) (*Glossary, error) {
+	glossary := Glossary{}
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(data, &glossary); err != nil {
+		return nil, err
+	}
+	return &glossary, nil
+}
+
+func buildWordMap(glossary *Glossary) (map[string]*GlossaryEntry, error) {
+	result := map[string]*GlossaryEntry{}
+	for ix := range glossary.Items {
+		entry := &glossary.Items[ix]
+		for _, word := range entry.Words {
+			_, exists := result[word]
+			if exists {
+				return nil, fmt.Errorf("Duplicate word: %s", word)
+			}
+			result[word] = entry
+		}
+	}
+	return result, nil
+}
+
+func indexOf(source, substr string, startingIndex int) int {
+	ix := strings.Index(source[startingIndex:], substr)
+	if ix == -1 {
+		return ix
+	}
+	return ix + startingIndex
+}
+
+func main() {
+	c := Config{}
+	pflag.StringVar(&c.inputFile, "input-file", "", "The input file to process, '-' for stdin")
+	pflag.StringVar(&c.outputFile, "output-file", "-", "The output file to process, '-' for stdout")
+	pflag.StringVar(&c.glossaryFile, "glossary", "", "The glossary file to use.")
+
+	pflag.Parse()
+
+	if len(c.inputFile) == 0 {
+		glog.Fatalf("--input-file is required.")
+	}
+	if len(c.outputFile) == 0 {
+		glog.Fatal("--output-file is required.")
+	}
+	if len(c.glossaryFile) == 0 {
+		glog.Fatal("--glossary is required.")
+	}
+
+	glossary, err := loadGlossary(c.glossaryFile)
+	if err != nil {
+		glog.Fatalf("Couldn't load glossary: %v", err)
+	}
+	wordMap, err := buildWordMap(glossary)
+	if err != nil {
+		glog.Fatalf("Couldn't build word map: %v", err)
+	}
+	var reader io.Reader
+	if c.inputFile == "-" {
+		reader = os.Stdin
+	} else {
+		var err error
+		reader, err = os.Open(c.inputFile)
+		if err != nil {
+			glog.Fatalf("Can't open %s: %v", c.inputFile, err)
+		}
+	}
+
+	var writer *os.File
+	if c.outputFile == "-" {
+		writer = os.Stdout
+	} else {
+		var err error
+		writer, err = os.Open(c.outputFile)
+		if err != nil {
+			glog.Fatalf("Can't open %s for writing: %v", c.outputFile, err)
+		}
+		writer.Truncate(0)
+	}
+
+	scanner := bufio.NewScanner(reader)
+	inCode := false
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "```") {
+			inCode = !inCode
+		}
+		if !inCode {
+			words := strings.Split(line, " ")
+			ix := 0
+			for _, word := range words {
+				word := strings.TrimSpace(word)
+				if entry, exists := wordMap[word]; exists {
+					ix = indexOf(line, word, ix)
+					// Skip already linkified or code blocks
+					if ix == 0 || (line[ix-1] != '[' && line[ix-1] != '`') {
+						line = line[0:ix] + fmt.Sprintf("[%s](%s)", word, entry.Link) + line[ix+len(word):]
```

Comment:
This makes the markdown very verbose when we start marking up every occurrence. Could we use reference instead of inline links?

e.g.
Markdown: 
```
[Services][] consist of [Pods][]!

// at the bottom of the document instead of in the middle of all the text
[services]: <https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md> "Kubernetes Services"
[pods]: <https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/pods.md> "Kubernetes Pods"
```

[Services][] consist of [Pods][]!

[services]: <https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md> "Kubernetes Services"
[pods]: <https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/pods.md> "Kubernetes Pods"